analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.ExceptionCandidate.cs

```diff
@@ -0,0 +1,516 @@
+        {
+            const string code = @"
+var tag = ""BeforeTry"";
+try
+{
+    tag = ""InTry"";
+    var x = p.firstName;
+}
+catch
+{
+    tag = ""InCatch"";
+}
+tag = ""AfterCatch"";";
+            var validator = SETestContext.CreateCS(code, ", Person p").Validator;
+            validator.ValidateTagOrder("BeforeTry", "InTry", "InCatch", "AfterCatch");
+            validator.TagStates("InCatch").Should().ContainSingle(x => HasUnknownException(x));
+            validator.ExitStates.Should().ContainSingle(x => HasNoException(x));
+        }
+
+        [DataTestMethod]
+        [DataRow("this.field")]
+        [DataRow("field")]
+        public void ExceptionCandidate_FieldReference_This(string field)
+        {
+            var code = $@"
+var tag = ""BeforeTry"";
+try
+{{
+    tag = ""InTry"";
+    var x = {field};
+}}
+catch
+{{
+    tag = ""InCatch"";
+}}
+tag = ""AfterCatch"";";
+            var validator = SETestContext.CreateCS(code).Validator;
+            validator.ValidateTagOrder("BeforeTry", "InTry", "AfterCatch");
+            validator.ExitStates.Should().ContainSingle(x => HasNoException(x));
+        }
+
+        [TestMethod]
+        public void ExceptionCandidate_FieldReference_Static()
+        {
+            const string code = @"
+var tag = ""BeforeTry"";
+try
+{
+    tag = ""InTry"";
+    var x = DateTime.MaxValue;
+}
+catch
+{
+    tag = ""InCatch"";
+}
+tag = ""AfterCatch"";";
+            var validator = SETestContext.CreateCS(code).Validator;
+            validator.ValidateTagOrder("BeforeTry", "InTry", "AfterCatch");
+            validator.ExitStates.Should().HaveCount(1).And.ContainSingle(x => HasNoException(x));
+        }
+
+        [TestMethod]
+        public void ExceptionCandidate_ForEachLoop()
+        {
+            const string code = @"
+var tag = ""BeforeTry"";
+try
+{
+    tag = ""InTry"";
+    foreach (string element in collection) // can throw if the collection is null or if it gets modified
+    {
+    }
+}
+catch
+{
+    tag = ""InCatch"";
+}
+tag = ""AfterCatch"";";
+            var validator = SETestContext.CreateCS(code, ", IEnumerable<string> collection").Validator;
+            validator.ValidateTagOrder("BeforeTry", "InTry", "InCatch", "AfterCatch", "InCatch", "AfterCatch");
+
+            // IForEachLoopOperation is not generated. It doesn't seem to be used.
+
+            // In the case of foreach, there are implicit method calls that in the current implementation can throw:
+            // - IEnumerable<>.GetEnumerator()
+            // - System.Collections.IEnumerator.MoveNext()
+            // - System.IDisposable.Dispose()
+            validator.TagStates("InCatch").Should().HaveCount(2).And.OnlyContain(x => HasUnknownException(x));
+            HasUnknownException(validator.ExitStates[0]).Should().BeTrue();
+            HasNoException(validator.ExitStates[1]).Should().BeTrue();
+            HasNoException(validator.ExitStates[2]).Should().BeTrue();
```

Comment:
The exit state order is not important and can change in the future. What about something like this?
```
validator.ExitStates.Should().HaveCount(3)
                .And.ContainSingle(x => true)
                .And.Subject.Where(HasNoException).Should().HaveCount(2);
```

Reply:
We might come up with a better overload for this in the other PR. Something that would take a predicate and expected count.

Reply:
This changes a bit in the follow-up PR. Let's to the change there.